### ISSUE TITLE
Menu size and graphical issues

### ISSUE THREAD

**Issue by red-kangaroo** [Attnam/ivan#368]
There are several small bugs with the new graphical options.

Option for size of menus seems to not work correctly. When you select larger amount of items per page, menus can overflow downwards, completely obscuring the text field. That is mostly a problem in shops, where some information is given not in prompts, but in normal messages.

![bug2](https://user-images.githubusercontent.com/11293143/40627297-b4c13e6a-62ad-11e8-938f-bbd265d986b9.png)

However, selecting smaller amount of items per menu page causes inventory menu to shrink to only a few items per page, which makes reviewing your inventory inconvenient. Drop and similar menus (all with alphabetic notation of items) will still overflow, though not so much.

![bug5](https://user-images.githubusercontent.com/11293143/40627445-95f3f576-62ae-11e8-8f9f-74d9d2b7029f.png)
![bug7](https://user-images.githubusercontent.com/11293143/40627449-9d917fba-62ae-11e8-8996-33077879ba97.png)

I was unable to find an option to only fill the base game window as the game did before, nor one that would make all menu types behave consistently.

With large amounts of items, the screen does not update correctly, leaving artifacts:
![bug](https://user-images.githubusercontent.com/11293143/40626929-381b6bb2-62ab-11e8-8dd3-40c74fc50e7f.png)
![bug4](https://user-images.githubusercontent.com/11293143/40627260-8be9d812-62ad-11e8-97e3-2f891a13b42a.png)

This is unlikely to happen in normal game, though, as it requires huge amounts of items on a single tile.

The list of items under your feet is plain awesome.

**Comment by AquariusPower** [comment]
The show items under is overflowing to the right of the bitmap, and then being drawn from the left again.
The squares above and below are ignored when sending a strong new draw request, therefore the glitches.

If I am not wrong, you are using vanilla dungeon right (scale x1)?
**I think the fast solution (bug fix) is to just limit the amount of shown items above head to the width of the dungeon in squares, so not all will be shown for now (as the glitch also messes their visualization anyways).**
A better but more complicated option (a true improvement) would be to show them in many lines ex.: for 20 items would show like 5x4 squares above player's head, filled with items gfx, and that must also be within the dungeon area, so it means a some more calculations.

**Comment by AquariusPower** [comment]
>I was unable to find an option to only fill the base game window as the game did before, nor one that would make all menu types behave consistently.

yep.. the list is a quite complicated thing, we cant set it's size based on pixels.

the only way is to limit the amount of entries.

as I remember, to worsen things, they are categorized, so the added categories (w/o icons), makes it even harder to put in the height limit, as the entries limit doesnt consider the categories entries!

That's why I left that height under user total control, to find the "best possible/usable" fit, so for now it just needs to be usable, to look really good will be an improvement.

if someone get a good idea to make the height calculation automatic to make it look really good, it will help a lot :)

May be it could process the list in two steps:
- 1st it would not draw, but prepare the list completely to find out the final height in pixels and then calculate the **precise entries limit**
- 2nd it would get the **precise entries limit** and will actually draw the list using it.
But that sounds... messy... 

Messier would be to make all the list calculations in double just to guess the **precise entries limit**, but then duplicated maintenance would be complicated.

So, may be, split the list creation in calculation and drawing could be the trick.

Unless I am mistaken in some of the conclusions here :)

**Comment by red-kangaroo** [comment]
One other issue related to lists/menus that I noticed:

When you use the F1 help button and then exit the help sub-menu, sometimes it doesn't properly disappear. This is most likely because the menu is redrawn as you return to it, but redrawing the whole map is not called. Thus when the menu is large enough, it is redrawn over the help sub-menu and the help properly disappears, but when the menu is small, the help will remain below the redrawn menu and results in artifacts that persist until you exit the menu altogether, when redrawing the whole screen is called.

**Comment by AquariusPower** [comment]
That's probably what is happening!

If I remember well, there are animations playing while the menu is opened right? like glowing wall lanterns, but, if I am not wrong, that wont happen if you scaled the dungeon x2 or more.

What needs a refresh in this case is the scaled dungeon, just after closing the help and before drawing the related list again, I think such call (to redraw the scaled up graphics) must be placed on the list code cpp file.

The main problem (and feature) is that the dungeon is scaled up in whole, not square by square. Scaling in whole provides better surrounding squares merging thru xBRZ (overall better looks), but prevents the optimizations that provides animations while the menus are opened (that would require square by square scaling up, and I think both codes could co-exist whenever the new code is implemented :)).

Obs.:
There is another problem, calling the help during "map notes" action, is buggy related to the **input**. If i remember well, after pressing F1, the help wont show up, and you have to press any key to let it show up (ex. space). It is a minor issue, and got too complicated (to me) to workaround while I was reading the code, so I gave up to do other stuff at that time.
